### PR TITLE
Ensure the_console_services() is constructed only once. 

### PR DESCRIPTION
**Pull request by RAOF**
We need to pass the_console_services() through to a number of components, some of which (like platform probing) don't have anywhere to stash an owning reference. This means we construct and destroy the_console_services() multiple times during server startup.

So, use a shared_ptr<> rather than a CachedPtr<> for console_services, leaving DefaultServerConfiguration to have an owning ref and guaranteeing single-construction.

This exposes a bug in `LogindConsoleServices`; the `Device` destructor did not actually release the logind device resulting in further attempts to acquire the device to error with an “already acquired” error; fixed by actually calling `ReleaseDevice` on `Device` destruction.

*That* exposed a deadlock in `GLibMainLoop::run_with_context_as_thread_default()` when called from the mainloop; fixed by adding an eager check for main context acquisition.

### CODE REVIEW

File: src/server/glib_main_loop.cpp [hunk 1]
```diff
@@ -398,51 +398,72 @@ void mir::GLibMainLoop::reprocess_all_sources()
...
-            {
-                promise->set_exception(std::current_exception());
-            }
             g_main_context_pop_thread_default(main_context);
             g_main_context_release(main_context);
-        });
+            throw;
+        }
+        g_main_context_pop_thread_default(main_context);
+        g_main_context_release(main_context);
```
Comment (RAOF) [inline src/server/glib_main_loop.cpp]: This is very nearly `mir::raii`able, but not quite, and it didn't seem worth the effort of doing something custom.

**Comment by RAOF** [comment]
GitHub is doing something weird with #502, so here it is again.

**Comment by RAOF** [comment]
bors try

**Comment by AlanGriffiths** [comment]
Seems functional, having a close look at the code...

File: src/server/console/default_configuration.cpp [hunk 2]
```diff
@@ -200,5 +200,13 @@ std::shared_ptr<mir::ConsoleServices> mir::DefaultServerConfiguration::the_conso
             BOOST_THROW_EXCEPTION((
                 std::runtime_error{
                     std::string{"Unknown console provider: "} + provider}));
-        });
+        };
+
+    // TODO: this is clearly not threadsafe, but then neither is CachedPtr
```
Comment (AlanGriffiths) [inline src/server/console/default_configuration.cpp]: That doesn't need to be called out here. *All* this "server configuration" code is intended to be used only during the single threaded initialization phase: prior to starting the server (and the associated threads).

**Comment by AlanGriffiths** [comment]
```
OCI runtime exec failed: exec failed: container_linux.go:348: starting container process caused "process_linux.go:86: executing setns process caused \"exit status 21\"": unknown
```

bors r=gerboland

**Comment by AlanGriffiths** [comment]
bors r=gerboland

**Pull request merged**